datafusion/optimizer/src/simplify_expressions.rs

```diff
@@ -1263,6 +1312,37 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_simplify_concat() {
+        fn build_concat_expr(args: &[Expr]) -> Expr {
+            Expr::ScalarFunction {
+                fun: BuiltinScalarFunction::Concat,
+                args: args.to_vec(),
+            }
+        }
+
+        let null = Expr::Literal(ScalarValue::Utf8(None));
+        // concat(true, c0, false, null, 'hello', c1, 12, 3.4) --> concat('1', c0, '0hello', c1, '123.4')
```

Comment:
For example, in this case the true, false, etc will already be cast to utf8 by the time simplify sees this:

```
Plan("'datafusion.public.values' not found")
❯ create table foo as values (1);
0 rows in set. Query took 0.000 seconds.
❯ select * from foo
;
+---------+
| column1 |
+---------+
| 1       |
+---------+
1 row in set. Query took 0.001 seconds.
❯ explain select concat(true, false, 'bar', column1) from foo;
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                 |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: concat(Boolean(true), Boolean(false), Utf8("bar"), #foo.column1)                                                                                         |
|               |   TableScan: foo projection=[column1]                                                                                                                                |
| physical_plan | ProjectionExec: expr=[concat(CAST(true AS Utf8), CAST(false AS Utf8), bar, CAST(column1@0 AS Utf8)) as concat(Boolean(true),Boolean(false),Utf8("bar"),foo.column1)] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                                                                                      |
|               |                                                                                                                                                                      |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
2 rows in set. Query took 0.002 seconds.
```

Reply:
Should we file another issue to add type-coercion rules for `concat` and `concat_ws`?